runtime/domain.c

```diff
@@ -169,18 +174,47 @@ asize_t caml_norm_minor_heap_size (intnat wsize)
 int caml_reallocate_minor_heap(asize_t wsize)
 {
   caml_domain_state* domain_state = Caml_state;
-  Assert(domain_state->young_ptr == domain_state->young_end);
+  uintnat global_minor_heap_ptr;
+  uintnat new_alloc_ptr;
+
+  caml_ev_begin("reallocate");
+
+  while (1) {
+
+    global_minor_heap_ptr = atomic_load_explicit(&caml_global_minor_heap_ptr, memory_order_acquire);
+
+    #ifdef DEBUG
+    CAMLassert(global_minor_heap_ptr != 0x0);
+    #endif
+
+    new_alloc_ptr = global_minor_heap_ptr + Bsize_wsize(wsize);
+
+    if (new_alloc_ptr > caml_minor_heaps_end)
+      return REALLOCATE_HEAP_FULL;
+
+    /* CAS away and bump the allocation pointer: if it fails a domain likely
+       snatched the requested segment. Restart the loop, else break. */
+    if(atomic_compare_exchange_strong(&caml_global_minor_heap_ptr,
+				      &global_minor_heap_ptr,
+				      new_alloc_ptr)) {
+      break;
+    };
+
+  }
 
   /* free old minor heap.
      instead of unmapping the heap, we decommit it, so there's
      no race whereby other code could attempt to reuse the memory. */
   caml_mem_decommit((void*)domain_self->minor_heap_area,
                     domain_self->minor_heap_area_end - domain_self->minor_heap_area);
```

Comment:
I think this works out. This borrows the same logic as the current reallocate code, which is that a segment is decommited before reallocation.
Before it worked because there was full ownership of a specific minor heap segment. (one segment per domain, always the same).

In this branch, every segments that one domain gets is in the rightful position to decommit the memory: everything beyond the global allocation pointer is unused memory (because resetting the global allocation always involves a minor collection).
So everything we decommit has supposedly been collected in the previous minor cycle (and values promoted if need be).
I don't think there is a path where we would touch values in `to-be-decommited` (beyond the allocation pointer) areas, but I may be missing a subtlety

As for the `REALLOCATE_HEAP_FULL`, hitting that codepath means there is no segment available, so there is nothing to decommit.

I think the confusing part of this logic is, do we actually need to decommit every segment we fetch?
I kept it the same as our current branch (+ accounting for the fact that there is no ownership of individual minor heaps segments per domains). The current code mentions this:
https://github.com/ocaml-multicore/ocaml-multicore/blob/parallel_minor_gc/runtime/domain.c#L174

```
 /* free old minor heap.
     instead of unmapping the heap, we decommit it, so there's
     no race whereby other code could attempt to reuse the memory. */
```

I'm slightly confused by this comment, is there a scenario where other code could attempt to reuse the memory, after performing a minor collection? (which is the pre-requisite for resetting the global allocation pointer)

Reply:
(or maybe it is a concurrent minor quirk?)